src/utils/AdaptiveTimeStep.cpp

```diff
@@ -46,55 +51,53 @@ AdaptiveTimeStep::AdaptiveTimeStep (const int nbeams)
 
 #ifdef AMREX_USE_MPI
 void
-AdaptiveTimeStep::NotifyTimeStep (amrex::Real dt, MPI_Comm a_comm_z)
+AdaptiveTimeStep::BroadcastTimeStep (amrex::Real& dt, MPI_Comm a_comm_z, int a_numprocs_z)
 {
     if (m_do_adaptive_time_step == 0) return;
-    const int my_rank_z = amrex::ParallelDescriptor::MyProc();
-    if (my_rank_z >= 1)
-    {
-        MPI_Send(&dt, 1, amrex::ParallelDescriptor::Mpi_typemap<amrex::Real>::type(),
-                 my_rank_z-1, comm_z_tag, a_comm_z);
-    }
-}
 
-void
-AdaptiveTimeStep::WaitTimeStep (amrex::Real& dt, MPI_Comm a_comm_z)
-{
-    if (m_do_adaptive_time_step == 0) return;
-    const int my_rank_z = amrex::ParallelDescriptor::MyProc();
-    if (!Hipace::HeadRank())
-    {
-        MPI_Status status;
-        MPI_Recv(&dt, 1, amrex::ParallelDescriptor::Mpi_typemap<amrex::Real>::type(),
-                 my_rank_z+1, comm_z_tag, a_comm_z, &status);
-    }
+    MPI_Bcast(&dt, 1, amrex::ParallelDescriptor::Mpi_typemap<amrex::Real>::type(),
+              a_numprocs_z - 1, a_comm_z);
+    MPI_Bcast(&m_min_uz, 1, amrex::ParallelDescriptor::Mpi_typemap<amrex::Real>::type(),
+              a_numprocs_z - 1, a_comm_z);
 }
+
 #endif
 
 void
-AdaptiveTimeStep::Calculate (amrex::Real& dt, MultiBeam& beams, amrex::Real plasma_density,
-                             const int it, const amrex::Vector<BoxSorter>& a_box_sorter_vec,
-                             const bool initial)
+AdaptiveTimeStep::Calculate (
+    amrex::Real t, amrex::Real& dt, MultiBeam& beams, MultiPlasma& plasmas,
+    const amrex::Geometry& geom, const Fields& fields,
+    const int it, const amrex::Vector<BoxSorter>& a_box_sorter_vec,
+    const bool initial)
 {
     HIPACE_PROFILE("AdaptiveTimeStep::Calculate()");
     using namespace amrex::literals;
 
     if (m_do_adaptive_time_step == 0) return;
-    if (!Hipace::HeadRank() && initial) return;
```

Comment:
With this line, I could not have proper calculation of the first time step of each rank. Since this line is removed, there are a few new `break` and `if` statements in the code, because the main problem is still present: most ranks don't know the beam at init.